### ISSUE TITLE
[Bug]: the scale of precision of result for vector type is non-deterministic

### ISSUE THREAD

**Issue by aressu1985** [matrixorigin/matrixone#12420]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Environment

```Markdown
- Version or commit-id (e.g. v0.1.0 or 8b23a93):4e4d1cd448867ac0fba7afcf53b576f72cd5ff30
- Hardware parameters:
- OS type:
- Others:
```

### Actual Behavior

1. mysql> select * from vtab32;
+------+------------------------------------+--------------------------------------------------------+
| id   | vecf32_3                           | vecf32_5                                               |
+------+------------------------------------+--------------------------------------------------------+
|    1 | NULL                               | NULL                                                   |
|    2 | [0.8166459, **0.66616553**, 0.4886152] | NULL                                                   |
|    3 | [0.1726299, 3.2908857, 30.433094]  | [0.45052445, 2.1984527, 9.579752, 123.48039, 4635.894] |
|    4 | [8.560689, 6.790359, 821.9778]     | [0.46323407, 23.498016, 563.923, 56.076736, 8732.958]  |
+------+------------------------------------+--------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> select id,vecf32_3 + vecf32_3, vecf32_5 + vecf32_5 from vtab32;
+------+-----------------------------------+---------------------------------------------------------+
| id   | vecf32_3 + vecf32_3               | vecf32_5 + vecf32_5                                     |
+------+-----------------------------------+---------------------------------------------------------+
|    1 | NULL                              | NULL                                                    |
|    2 | [1.6332918, **1.3323311**, 0.9772304] | NULL                                                    |
|    3 | [0.3452598, 6.5817714, 60.866188] | [0.9010489, 4.3969054, 19.159504, 246.96078, 9271.788]  |
|    4 | [17.121378, 13.580718, 1643.9556] | [0.92646813, 46.996033, 1127.846, 112.15347, 17465.916] |
+------+-----------------------------------+---------------------------------------------------------+

the result should be **1.33233106, not 1.3323311**

2.mysql> select id,vecf32_3 * vecf32_3, vecf32_5 * vecf32_5 from vtab32;
+------+-------------------------------------+--------------------------------------------------------------+
| id   | vecf32_3 * vecf32_3                 | vecf32_5 * vecf32_5                                          |
+------+-------------------------------------+--------------------------------------------------------------+
|    1 | NULL                                | NULL                                                         |
|    2 | [0.6669106, 0.44377652, 0.23874483] | NULL                                                         |
|    3 | [0.02980108, 10.829928, 926.1732]   | [0.20297228, 4.8331943, 91.771645, 15247.407, **2.1491514e+07**] |
|    4 | [73.28539, 46.108974, 675647.5]     | [0.2145858, 552.1568, 318009.12, 3144.6003, **7.626455e+07**]    |
+------+-------------------------------------+--------------------------------------------------------------+

On which condition, the result is the format of scientific notation?

3.mysql> select l2_norm(vecf64_3), l2_norm(vecf64_5) from vtab64;
+--------------------+--------------------+
| l2_norm(vecf64_3)  | l2_norm(vecf64_5)  |
+--------------------+--------------------+
|               NULL |               NULL |
| 1.1616504866000061 |               NULL |
|  **822.0503910711226** |  **8751.358081384826** |
| **0.9657884379576713** | **339.69107269209786** |
+--------------------+--------------------+

why this precision scale is so large?

**Comment by arjunsk** [comment]
Expected behavior is `0.66616553 + 0.66616553 = 1.33233106`

After debugging, I found that, for

```sql
mysql> select 0.66616553+0.66616553;
+-------------------------+
| 0.66616553 + 0.66616553 |
+-------------------------+
|              1.33233106 |
+-------------------------+
```

we treat "0.66616553" as T_decimal64 by default rather than T_float64. 

For it to be treated as T_float32, we can cast one of the arguments
```sql
mysql> select cast(0.66616553 as float)+0.66616553;
+----------------------------------------+
| cast(0.66616553 as float) + 0.66616553 |
+----------------------------------------+
|                              1.3323311 |
+----------------------------------------+

```

I found a bug in the VectorToString conversion code. I will be fixing it.

**Comment by arjunsk** [comment]
This PR should resolve this issue: https://github.com/matrixorigin/matrixone/pull/12428

1. the result should be 1.33233106, not 1.3323311

The output should 1.3323311 for float32 type.
```sql
+-------------------------------------------------------+
| cast(0.66616553 as float) + cast(0.66616553 as float) |
+-------------------------------------------------------+
|                                             1.3323311 |
+-------------------------------------------------------+

```

2. On which condition, the result is the format of scientific notation?

That was a bug. Ideally, it should print
```sql
mysql> select a, b * b, c * c from t7;
+------+-------------------------------------+---------------------------------------------------------+
| a    | b * b                               | c * c                                                   |
+------+-------------------------------------+---------------------------------------------------------+
|    1 | NULL                                | NULL                                                    |
|    2 | [0.6669106, 0.44377652, 0.23874483] | NULL                                                    |
|    3 | [0.02980108, 10.829928, 926.1732]   | [0.20297228, 4.8331943, 91.771645, 15247.407, 21491514] |
|    4 | [73.28539, 46.108974, 675647.5]     | [0.2145858, 552.1568, 318009.12, 3144.6003, 76264550]   |
+------+-------------------------------------+---------------------------------------------------------+

```
3. The precision scale is large because the return type of l2_norm is double. Double has a similar scale.

```sql
mysql> select cast(339.69107269209786 as double);
+------------------------------------+
| cast(339.69107269209786 as double) |
+------------------------------------+
|                 339.69107269209786 |
+------------------------------------+

mysql> select cast(339.69107269209786 as float);
+-----------------------------------+
| cast(339.69107269209786 as float) |
+-----------------------------------+
|                         339.69107 |
+-----------------------------------+

```

**Comment by arjunsk** [comment]
Waiting for @aunjgr to approve the PR. I will update the thread once the PR is merged into the master.

**Comment by arjunsk** [comment]
PR is merged into the master. Kindly retest.

**Comment by aressu1985** [comment]
fixed

**Comment by arjunsk** [comment]
Output for double, float and decimal.

```sql
--- 1.a implicit double [VECF64]
select 8732.958 * 8732.958;

+---------------------+
| 8732.958 * 8732.958 |
+---------------------+
|     76264555.429764 |
+---------------------+


--- 2.a explicit double (same as implicit double.) [VECF64]
select cast(8732.958 as double) * cast(8732.958 as double);
+-----------------------------------------------------+
| cast(8732.958 as double) * cast(8732.958 as double) |
+-----------------------------------------------------+
|                                     76264555.429764 |
+-----------------------------------------------------+

--- 3.a float (will loose precision)
select cast(8732.958 as float) * cast(8732.958 as float); [VECF32]
+---------------------------------------------------+
| cast(8732.958 as float) * cast(8732.958 as float) |
+---------------------------------------------------+
|                                          76264550 |
+---------------------------------------------------+

--- 4.a decimal (used when you want the exact decimal value part without any rounding) [Decimal is not supported in VECTOR]
select cast(8732.958 as decimal(10,5)) * cast(8732.958 as  decimal(10,5));
+---------------------------------------------------------------------+
| cast(8732.958 as decimal(10, 5)) * cast(8732.958 as decimal(10, 5)) |
+---------------------------------------------------------------------+
|                                                 76264555.4297640000 |
+---------------------------------------------------------------------+
```

For vecfxx

```sql
--- 1.b vecf64 (equivalent to double)
select cast("[8732.958]" as vecf64(1)) * cast("[8732.958]" as vecf64(1));

+---------------------------------------------------------------+
| cast([8732.958] as vecf64(1)) * cast([8732.958] as vecf64(1)) |
+---------------------------------------------------------------+
| [76264555.429764]                                             |
+---------------------------------------------------------------+

--- 3.b vecf32 (equivalent to float)
select cast("[8732.958]" as vecf32(1)) * cast("[8732.958]" as vecf32(1));

+---------------------------------------------------------------+
| cast([8732.958] as vecf32(1)) * cast([8732.958] as vecf32(1)) |
+---------------------------------------------------------------+
| [76264550]                                                    |
+---------------------------------------------------------------+
```

**Comment by aressu1985** [comment]
testing

**Comment by aressu1985** [comment]
fixed
